alerts/HighlyAvailableWorkloadIncorrectlySpread.md

```diff
@@ -0,0 +1,81 @@
+# HighlyAvailableWorkloadIncorrectlySpread
+
+## Meaning
+
+Highly available workloads with persistent storage are incorrectly spread across
+multiple nodes. In high availability topology, all workloads with multiple
+replicas should be spread across multiple nodes to avoid having a single point
+of failure. However, workload with persistent storage are more complex to
+schedule since persistent volumes may be bound to an availability zone, so they
+need to be spread manually before any scheduling constraints are put in place.
+
+## Impact
+
+The cluster isn't fully highly-available since workloads with persistent
+storage might still have a single point of failure.
+
+## Diagnosis
+
+The alert indicates which workloads aren't correctly spread across multiple
+nodes. You can find information to identify them under the `workload` and
+`namespace` labels.
+
+```console
+ - alertname = HighlyAvailableWorkloadIncorrectlySpread
+...
+ - workload = prometheus-k8s
+...
+ - namespace = openshift-monitoring
+```
+
+Then, you can verify that multiple instances of the same application are indeed
+scheduled on the same node by running:
+
+```console
+$ oc -n "$NS" get -o wide pods | grep "$WORKLOAD"
+```
+
+## Mitigation
+
+The script below is provided to rebalance workloads across multiple nodes. To
+use it, you should first get the following information from the alert labels:
+* `workload`: workload name
+* `namespace`: namespace of the workload
+* `node`: node hosting multiple instances of the workload
+
+Then you will want to update the script to include the workloads that need to be
+rebalance across multiple nodes. To do so you can add the following call for
+each workload reported by the alert at the end of the script below:
+
+```sh
+rebalancePods $WORKLOAD $NS $NODE
+```
+
+For example if you need to rebalance `prometheus-k8s`, you would have to add the
+following line at the end of the script:
+```sh
+rebalancePods prometheus-k8s openshift-monitoring ip-10-0-0-0.ec2.internal
+```
+
+```sh
+#!/bin/bash
+
+AZ_COUNT=$(oc get nodes -o yaml | grep -E "^\s+failure-domain.beta.kubernetes.io/zone" | uniq | wc -l)
+
+rebalancePods() {
+    WORKLOAD=$1
+    NS=$2
+    NODE=$3
+
+    oc adm cordon "$NODE"
+    PODS=$(oc -n "$NS" get -o wide pods | grep "$WORKLOAD.*$NODE" | cut -f1 -d ' ' | head -n-1)
+    for POD in $PODS; do
+        if [ "${AZ_COUNT}" -gt "1" ]; then
+            PVC=$(oc get pod -n "$NS" "$POD" -ojson | jq -r '.spec.volumes[] | select(.persistentVolumeClaim!=null) | .persistentVolumeClaim.claimName')
+            oc delete pvc --wait=false -n "$NS" "$PVC"
+        fi
+        oc delete pod -n "$NS" "$POD"
+    done
+    oc adm uncordon "$NODE"
+}
+```
```

Comment:
Personally, I wouldn't put a script in a runbook. Usually, mitigation commands are executed directly as-is by the administrator (frequently by copy-paste), so making it short is important.

Reply:
my biggest concern with this would be customers asking "well if you know what to do, why doesn't the product do it"

Reply:
yes good point rick, that was also my concern when adding this script. I'll convert it to make it look like steps for the users to follow in order to fix the issue

Reply:
I changed the **Mitigation** part according to your comments, please have a look :)